packages/transactions/tests/structuredDataSignature.test.ts

```diff
@@ -0,0 +1,190 @@
+      recipient: standardPrincipalCV(principal1),
+      salt: uintCV('12345'),
+    });
+    const domain1 = 'string example' as any; // Cast the type to trigger the error
+
+    expect(() => hashStructuredData(domain1)).toThrowError();
+    expect(() => hashStructuredData(message1)).not.toThrowError(); // Cast the type to throw error
+    expect(() => encodeStructuredData(message1, domain1)).toThrowError();
+  });
+
+  test('encoding / decoding failures bad message type', () => {
+    const message1 = 'another string message' as any; // Cast the type to trigger the error
+    const domain1 = tupleCV({
+      name: stringAsciiCV('hiro.so'),
+      version: stringAsciiCV('1.0.0'),
+      'chain-id': uintCV(chainIds['mainnet']),
+    });
+
+    expect(() => hashStructuredData(domain1)).not.toThrowError();
+    expect(() => hashStructuredData(message1)).toThrowError(); // Cast the type to throw error
+    expect(() => encodeStructuredData(message1, domain1)).toThrowError();
+  });
+
+  test('encoding different inputs produces different outputs', () => {
+    const input1 = {
+      message: tupleCV({
+        amount: uintCV('100'),
+        recipient: standardPrincipalCV(principal1),
+        salt: uintCV('12345'),
+      }),
+      domain: tupleCV({
+        name: stringAsciiCV('docs.stacks.co'),
+        version: stringAsciiCV('1.0.0'),
+        'chain-id': uintCV(chainIds['mainnet']),
+      }),
+    };
+
+    const input2 = {
+      message: tupleCV({
+        amount: uintCV('100'),
+        recipient: standardPrincipalCV(principal1),
+        salt: uintCV('12345'),
+      }),
+      domain: tupleCV({
+        name: stringAsciiCV('hiro.so'),
+        version: stringAsciiCV('1.0.0'),
+        'chain-id': uintCV(chainIds['mainnet']),
+      }),
+    };
+
+    const input1Encoded = encodeStructuredData(input1.domain, input1.message);
+    const input2Encoded = encodeStructuredData(input2.domain, input2.message);
+    expect(input1Encoded).not.toEqual(input2Encoded);
+  });
+});
+
+// See https://github.com/MarvinJanssen/sips/blob/feat/signed-structured-data/sips/sip-018/sip-018-signed-structured-data.md#test-vectors
+describe('SIP018 test vectors', () => {
+  test('Structured data hashing', () => {
+    const inputs: [ClarityValue, string][] = [
+      [
+        stringAsciiCV('Hello World'),
+        '5297eef9765c466d945ad1cb2c81b30b9fed6c165575dc9226e9edf78b8cd9e8',
+      ],
+      // CV =  (empty string):
+      [stringAsciiCV(''), '3c8f1b104592e3ebb2b2602b3979a27e77f586fb4c655369fa4eccb6d545a0f8'],
+      [
+        // CV =  (domain tuple):
+        tupleCV({
+          name: stringAsciiCV('Test App'),
+          version: stringAsciiCV('1.0.0'),
+          'chain-id': uintCV(1),
+        }),
+        '2538b5dc06c5ae2f11549261d7ae174d9f77a55a92b00f330884695497be5065',
+      ],
+    ];
+
+    for (const [input, expectedHash] of inputs) {
+      expect(hashStructuredData(input).toString('hex')).toEqual(expectedHash);
```

Comment:
jest also provides [parameterized testing](https://jestjs.io/docs/api#testeachtablename-fn-timeout) for these `for` tests.
_(also relevant for test `'encodeStructuredData / decodeStructuredDataSignature'`)_